ui/src/app/components/breadcrumb.tsx

```diff
@@ -27,75 +34,95 @@ const styles = {
   }
 };
 
+class BreadcrumbData {
+  label: string;
+  url: string;
+
+  constructor(label: string, url: string) {
+    this.label = label;
+    this.url = url;
+  }
+}
+
 // Generates a trail of breadcrumbs based on currently loaded data.
-export const getTrail = (type: BreadcrumbType, data): {label: string, url: string}[] => {
-  const {workspace, cohort, conceptSet, urlParams: {ns, wsid, cid, csid, pid, nbName}} = data;
+export const getTrail = (
+  type: BreadcrumbType,
+  workspace: WorkspaceData,
+  cohort: Cohort,
+  conceptSet: ConceptSet,
+  urlParams: any
+): Array<BreadcrumbData> => {
+  const {ns, wsid, cid, csid, pid, nbName} = urlParams;
   const prefix = `/workspaces/${ns}/${wsid}`;
   switch (type) {
     case BreadcrumbType.Workspaces:
       return [
-        {label: 'Workspaces', url: '/workspaces'}
+        new BreadcrumbData('Workspaces', '/workspaces')
       ];
     case BreadcrumbType.Workspace:
       return [
-        ...getTrail(BreadcrumbType.Workspaces, data),
-        {label: workspace ? workspace.name : '...', url: `${prefix}/data`}
+        ...getTrail(BreadcrumbType.Workspaces, workspace, cohort, conceptSet, urlParams),
```

Comment:
What do you mean?

Reply:
I've tried a few different things, including leaving the args as the original 'data' (compile time errors), changing it to a spread parameter (runtime errors), and defining consts outside the scope of this function that call this function (only slightly neater)

Reply:
Sorry, that wasn't clear at all. Not worth worrying about. I was wondering if this might be a place to try function [currying](https://wsvincent.com/javascript-currying/), but that would almost just be showing off. So a local function would capture the (local) arguments that are the same between all the calls, leaving only a one-arg function to operate on the remaining ones. No need to worry about it.